src/parser/ts/index.ts

```diff
@@ -0,0 +1,528 @@
+      for (const t of temp) {
+        if (locked) {
+          temptemp.push([...t, array]);
+        } else {
+          for (const item of array) {
+            temptemp.push([...t, [item]]);
+          }
+        }
+      }
+      res = [...temptemp];
+    } else {
+      if (locked) {
+        res = [[array]];
+      } else {
+        res = array.map(i => [[i]]);
+      }
+    }
+  }
+  return res;
+};
+
+export const convertExpressionToLabel = (
+  expression: string,
+  labelSrc: string,
+  tsMap: TSResponseMap
+) => {
+  const parsed = parse(expression);
+  const labelRes = reverse(parsed, item => {
+    if (isSTSReference(item)) {
+      const [{ value }] = getIdFilters(item);
+      const serie = tsMap[String(value)];
+      if (labelSrc) {
+        return getLabelWithInjectedProps(labelSrc, serie);
+      }
+      return `${serie.name || serie.externalId || serie.id}`;
+    }
+  });
+  return labelRes;
+};
+
+export const getLabelsForExpression = async (
+  expressions: string[],
+  labelSrc: string,
+  target,
+  connector: Connector
+) => {
+  const tsIds = getReferencedIdsInExpressions(expressions);
+  const tsUniqueIds = uniqBy(tsIds, unwrapId);
+  const referencedTS = await getTimeseries({ items: tsUniqueIds }, target, connector, false);
+  const tsMap = reduceTsToMap(referencedTS);
+  return expressions.map(expr => convertExpressionToLabel(expr, labelSrc, tsMap));
+};
+
+export const getReferencedTimeseries = (
+  route: STSQueryItem[] | STSFunction
+): StringMap[] => {
+  let idFilters: STSServerFilter[] = [];
+  walk(route, obj => {
+    if (isSTSReference(obj)) {
+      const ids = getIdFilters(obj);
+      idFilters = [...idFilters, ...ids];
+    }
+  });
+  return idFilters.map(({ path, value }) => ({
+    [path]: value as string,
+  }));
+};
+
+export const hasAggregates = (
+  expression: string
+): boolean => {
+  let hasAggregates = false;
+  walk(parse(expression), obj => {
+    if (isSTSReference(obj) && obj.query.some(isSTSAggregateFilter)) {
+      hasAggregates = true;
+    }
+  });
+  return hasAggregates
+};
+
+export function flattenServerQueryFilters(items: STSFilter[]): StringMap {
+  return items.reduce((res, filter) => {
+    let value: any;
+    if (isByIdsQuery(filter.value)) {
+      value = filter.value.map(item => flattenServerQueryFilters([item]));
+    } else if(isSTSFilterArr(filter.value)) {
+      value = {...res[filter.path], ...flattenServerQueryFilters(filter.value)};
+    } else if (isArray(filter.value)){
+      value = filter.value.map(flattenServerQueryFilters);
+    } else {
+      value = filter.value
+    }
+    res[filter.path] = value;
+    return res;
+  }, {});
+}
+
+export const walk = (
+  route: STSItem,
+  iterator: (item: STSQueryItem | STSFunction) => any
+) => {
+  if (isArray(route)) {
+    return route.map(r => walk(r, iterator));
+  }
+  const resArr = [];
+  const res = iterator(route);
+  if (res !== undefined) {
+    resArr.push(res);
+  }
+  if (isSTSFunction(route)) {
```

Comment:
can we have a case where both `if` statements will be `true` (I can't see)? If no then we can replace this logic with `if .. else` and get rid of `resArr`

Reply:
It doesn't seem like I ever use return functionality, I will just remove this chunk.

Reply:
fixed